pkg/antctl/raw/supportbundle/command_test.go

```diff
@@ -0,0 +1,346 @@
+			expectedClients: []string{"node-1", "node-3"},
+		},
+		{
+			name:            "failure to create one agent client due to error when parsing controller IP",
+			k8sClientset:    fake.NewSimpleClientset(&node2, &node3),
+			antreaClientset: fakeclientset.NewSimpleClientset(agentInfo1, agentInfo2),
+			expectedClients: []string{"node-3"},
+		},
+		{
+			name:            "no-ops due to error in listing antreaagentinfos",
+			antreaClientset: fakeclientset.NewSimpleClientset(),
+			prepareReactor1: func(clientset *fakeclientset.Clientset) {
+				clientset.PrependReactor("list", "antreaagentinfos", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, &v1beta1.AntreaAgentInfoList{}, errors.New("error listing antreaagentinfos")
+				})
+			},
+			expectedErr: "error listing antreaagentinfos",
+		},
+		{
+			name:            "no-ops due to error in listing nodes",
+			antreaClientset: fakeclientset.NewSimpleClientset(),
+			k8sClientset:    fake.NewSimpleClientset(),
+			prepareReactor2: func(clientset *fake.Clientset) {
+				clientset.PrependReactor("list", "nodes", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, &v1.NodeList{}, errors.New("error listing nodes")
+				})
+			},
+			expectedErr: "error listing nodes",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.prepareReactor1 != nil {
+				tt.prepareReactor1(tt.antreaClientset)
+			}
+			if tt.prepareReactor2 != nil {
+				tt.prepareReactor2(tt.k8sClientset)
+			}
+			clients, err := createAgentClients(tt.k8sClientset, tt.antreaClientset, clientConfig, "", nameList)
+			if tt.expectedErr != "" {
+				assert.ErrorContains(t, err, tt.expectedErr)
+			} else {
+				require.NoError(t, err)
+				agentClients := []string{}
+				for client := range clients {
+					agentClients = append(agentClients, client)
+				}
+				sort.Strings(agentClients)
+				assert.Equal(t, tt.expectedClients, agentClients)
+			}
+		})
+	}
+}
+
+func TestRequest(t *testing.T) {
+	k8sClientset := fake.NewSimpleClientset(&node1, &node3)
+	antreaClientset := fakeclientset.NewSimpleClientset(agentInfo1, agentInfo2, &controllerInfo)
+	agentClients, err := createAgentClients(k8sClientset, antreaClientset, clientConfig, "", nameList)
+	require.NoError(t, err)
+	controllerClient, err := createControllerClient(k8sClientset, antreaClientset, clientConfig)
+	require.NoError(t, err)
+	amount := len(agentClients)*2 + 2
+	bar := barTmpl.Start(amount)
+	defer bar.Finish()
+	defer bar.Set("prefix", "Finish ")
+	expectedResults := map[string]string{
+		"":       "Post \"http://172.16.0.0:0/supportbundle/system.antrea.io/v1beta1/supportbundles\": dial tcp 172.16.0.0:0:",
+		"node-1": "Post \"http://172.16.0.0:0/supportbundle/system.antrea.io/v1beta1/supportbundles\": dial tcp 172.16.0.0:0:",
+		"node-3": "Post \"http://127.0.0.1:0/supportbundle/system.antrea.io/v1beta1/supportbundles\": dial tcp 127.0.0.1:0:",
```

Comment:
Can you elaborate here? I have used fake antreaclientset and k8sclientset to create the restclients such as agentclients and controller client.....Is there any reference where I can see how to use the above package to create a fake rest client?

Reply:
The problem is not with the K8s clients, it is with the REST clients that are created in `createAgentClients` and `createControllerClient`. By calling these functions, you are using real clients which actually try to open an HTTP connection (which obviously fails because there is no server).

Test cases such as this one do not need to call `createAgentClients` and `createControllerClient`. These functions are not the ones being tested here. So just remove these calls and use an instead of a fake REST client instead. When you call `requestAll` (the function under test), use the fake clients.

For example, instead of:
```go
controllerClient, err := createControllerClient(k8sClientset, antreaClientset, clientConfig, ctx)
require.NoError(t, err)
```
Use:
```go
controllerClient:= fakerest.CreateHTTPClient(func(request *http.Request) (*http.Response, error) {
    return &http.Response{StatusCode: http.StatusOK, ...}
})
```

There are some examples in the client-go repo itself: https://github.com/search?q=repo%3Akubernetes%2Fclient-go+CreateHTTPClient&type=code